File: anim3d.cpp
#include <cocos2d.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

#include "anim3d.h"

void setAction(struct Obj *obj, const char *actname)
{
    int i = 0;
    obj->act = NULL;
    for(i = 0; i < obj->model->num_acts; i++)
    {
        struct Action *a = obj->model->acts+i;
        if(strcmp(actname, a->name) == 0)
        {
            obj->act = a;
            obj->cur_frame = a->kf[0].frame_num;
            break;
        }
    } 
    if(obj->act == NULL)
    {   
        obj->act = obj->model->acts;
    }
}

void renderObject(struct Obj *obj)
{
    int i = 0;
    int n_index = 0; 
    int index = 0;
    float *verts = NULL; 
    float *verts1 = NULL;
    float t = 0; //0 is first frame will be drawn.. 1 the second.. (keyframes)
    
    //3 coords per vertex and 3 vertices per face.
    float *drawv = (float*)malloc(sizeof(float) * obj->model->num_faces * 3 * 3);
    if(drawv == NULL)
    {
        DEBUG_PRINT(stderr, "Could not allocate memory");
        return;
    }
    
    int *indx = obj->model->indices;
    float *uv = obj->model->uv;
    
    index = 0;
    n_index = 1;
    
    while(obj->cur_frame > obj->act->kf[n_index].frame_num)
    {
        index++;
        n_index++;
        if(n_index == obj->act->num_frames)
        {
            index = 0;
            n_index = 1;
            obj->cur_frame = obj->act->kf[index].frame_num;
            break;
        }
    }


    verts = obj->act->kf[index].verts; //get the vertex from the first key frame
    verts1 = obj->act->kf[n_index].verts; //get the vertex from the first key frame

    t = (obj->cur_frame - obj->act->kf[index].frame_num)/(float)(obj->act->kf[n_index].frame_num - obj->act->kf[index].frame_num);//obj->cur_frame); 
    //DEBUG_PRINT(stderr, "%d %d %d %f\n", obj->cur_frame, index, n_index, t);
    
    int num_indices = obj->model->num_faces * 3; //3 indices per face.
    for(i = 0; i < num_indices; i++)
    {
        int v_index = indx[i];
        
        drawv[i*3 + 0] = verts[v_index*3 + 0] + t*(verts1[v_index*3 + 0]-verts[v_index*3 + 0]);
        
        drawv[i*3 + 1] = verts[v_index*3 + 1] + t*(verts1[v_index*3 + 1]-verts[v_index*3 + 1]);
        
        drawv[i*3 + 2] = verts[v_index*3 + 2] + t*(verts1[v_index*3 + 2]-verts[v_index*3 + 2]);
    }
    
    
    glVertexAttribPointer(cocos2d::kCCVertexAttrib_Position, 3, GL_FLOAT, GL_FALSE, 0, drawv);    
    glVertexAttribPointer(cocos2d::kCCVertexAttrib_TexCoords, 2, GL_FLOAT, GL_FALSE, 0, uv);
	glDrawArrays(GL_TRIANGLES, 0, obj->model->num_faces*3);
    
    free(drawv);
}


void freeModel(struct Model *m)
{
    int i, j;
    //free all the face information (indices, uv)
    free(m->indices);
    free(m->uv);
    
    for(i = 0; i < m->num_acts; i++)
    {
        for(j = 0; j < m->acts[i].num_frames; j++)
        {
            free(m->acts[i].kf[j].verts);
        }
        free(m->acts[i].kf);
    }
    free(m->acts);
}

struct Model * loadModel(const char *filename)
{
    FILE *fp = NULL;
    int face_offset = 0;
    int i = 0;
    int j = 0;
    int k = 0;
    fp = fopen(filename, "r");
    if(fp == NULL)
        exit(1);
    struct Model *m = (struct Model*) malloc(sizeof(struct Model));
    fscanf(fp, "%d", &(m->num_vert));
    fscanf(fp, "%d", &(m->num_faces));
    fscanf(fp, "%d", &(m->num_acts));
    //allocate memory for the faces. i.e, allocate memory for the indices
    m->indices = (int *) malloc(sizeof(int) * 3 * m->num_faces);    //3 indices per triangle
    m->uv = (float*) malloc(sizeof(float) * 2 * 3 * m->num_faces);  //2 uv per vertex(index) per triangle
    DEBUG_PRINT(stderr, "%d %d %d\n", m->num_vert, m->num_faces, m->num_acts);
    for(i = 0; i < m->num_faces; i++)
    {
        fscanf(fp, "%d %d %d", &m->indices[i*3+0], &m->indices[i*3+1], &m->indices[i*3+2]);
        DEBUG_PRINT(stderr, "%d %d %d\n", m->indices[i*3+0], m->indices[i*3+1], m->indices[i*3+2]);
        for(face_offset = i * 6, j = 0; j < 3; j++)
        {
            fscanf(fp, "%f %f", &m->uv[face_offset+j*2], &m->uv[face_offset+j*2+1]);
            DEBUG_PRINT(stderr, "%.3f %.3f\n", m->uv[face_offset+j*2], m->uv[face_offset+j*2+1]);
        }
    }
    //allocate memory for the actions
    m->acts = (struct Action*) malloc(sizeof(struct Action) * m->num_acts);
    for(i = 0; i < m->num_acts; i++)
    {
        fscanf(fp, "%s %d", m->acts[i].name, &m->acts[i].num_frames);    
        DEBUG_PRINT(stderr, "%s %d\n", m->acts[i].name, m->acts[i].num_frames);    
        ///allocate memory for the keyframes
        m->acts[i].kf = (struct KeyFrame *) malloc(sizeof(struct KeyFrame) * m->acts[i].num_frames);
        for(j = 0; j < m->acts[i].num_frames; j++)
        {
            float * verts = (float *) malloc(sizeof(float) * 3 * m->num_vert); 
            fscanf(fp, "%d", &m->acts[i].kf[j].frame_num);
            DEBUG_PRINT(stderr, "%d\n", m->acts[i].kf[j].frame_num);
            //m.acts[i].kf[j].verts //
            for(k = 0; k < m->num_vert; k++)
            {
                fscanf(fp, "%f %f %f", verts+k*3, verts+k*3+1, verts+k*3+2);   
                DEBUG_PRINT(stderr, "%.3f %.3f %.3f\n", verts[k*3], verts[k*3+1], verts[k*3+2]);   
            }
            m->acts[i].kf[j].verts = verts;
        }
        
    }
    fclose(fp);    
    return m;
}

File: loader/python/anim3d.py
from pyglet.gl import *
import math

class Face:
    def __init__(self, ind, uv):
        self.indices = ind 
        self.uv = uv

class KeyFrame:
    def __init__(self, fp, nv):
        self.verts = []
        self.frame_num = int(fp.readline())
        for i in xrange(nv):
            (v1, v2, v3) = fp.readline().split(' ')
            self.verts.append(float(v1)) 
            self.verts.append(float(v2)) 
            self.verts.append(float(v3)) 

class Action:
    def __init__(self, fp, nv):
        (self.name, self.num_frames) = fp.readline().split(' ')
        self.num_frames = int(self.num_frames)
        self.kf = []
        for j in xrange(self.num_frames):
            k = KeyFrame(fp, nv)
            self.kf.append(k)
    
#function to initialize values in a list.
def init_uv(v):
    return -1

class Model:
    def __init__(self, fname):
        fp = open(fname)
        (self.nv, self.nf, self.na) = fp.readline().split(' ')
        self.nv = int(self.nv)
        self.nf = int(self.nf)
        self.na = int(self.na)
        self.indices = ()
        self.uv = range(0, 2*self.nv)
        self.uv = map(init_uv, self.uv)
        self.repeats = ()
        #load the model data
        for i in xrange(self.nf):
            line = fp.readline().rstrip(' \n')
            indx = map(int, line.split(' '))
            for j in xrange(3):
                (s, t) = map(float, fp.readline().split(' '))
                if(self.uv[int(indx[j])*2] < 0):
                    self.uv[int(indx[j])*2] = s
                    self.uv[int(indx[j])*2+1] = t
                elif math.fabs(self.uv[int(indx[j])*2]-s) < 0.0001 and math.fabs(self.uv[int(indx[j])*2+1]-t) < 0.0001:
                    self.repeats += (int(indx[j]),) 
                    self.uv += [s, t]
                    indx[j] = self.nv + (len(self.repeats)-1)
            self.indices += (indx[0], indx[1], indx[2])
        #load the animation data
        self.acts = []
        for i in xrange(self.na):
            act = Action(fp, self.nv)
            self.acts.append(act)
        fp.close()

        print "Model: ", fname
        print "Faces: %d Anims: %d" % (self.nf, len(self.acts))
        for act in self.acts:
            print act.name

def linear_interop(v, v1, t):
    return v + t*(v1-v)
    
class AnimObject:
    def __init__(self):
        self.model = None
        self.act = None
        self.cur_frame = 0

    def loadModel(self, fname):
        pass
   
    def setModel(self, model):
        self.model = model

    def setAction(self, actname):
        self.act = None
        for i in xrange(self.model.na):
            a = self.model.acts[i]
            if(actname == a.name):
                self.act = a
                self.cur_frame = a.kf[0].frame_num
        if(self.act == None):
            self.act = self.obj.model.acts[0]

    def renderObject(self):
        if(self.cur_frame > self.act.kf[self.act.num_frames-1].frame_num):
            self.cur_frame = self.act.kf[0].frame_num+1

        n_index = 0
        while (self.cur_frame > self.act.kf[n_index].frame_num):
            n_index += 1
        index = n_index - 1

        v = self.act.kf[index].verts
        v1 = self.act.kf[n_index].verts
        t = (self.cur_frame - self.act.kf[index].frame_num)/float(self.act.kf[n_index].frame_num - self.act.kf[index].frame_num)
       
        dv = map(linear_interop, v, v1, (t, ) * len(v))

        for re in self.model.repeats:
            dv += (dv[re*3], dv[re*3+1], dv[re*3+2])

        pyglet.graphics.draw_indexed(self.model.nv + len(self.model.repeats), pyglet.gl.GL_TRIANGLES, self.model.indices, ('v3f', dv), ('t2f', self.model.uv))            
        
